hazelcast/src/main/java/com/hazelcast/instance/BuildInfo.java

```diff
@@ -59,6 +72,37 @@ public byte getSerializationVersion() {
         return serializationVersion;
     }
 
+    public static int calculateVersion(String version) {
+        if (null == version) {
+            return UNKNOWN_HAZELCAST_VERSION;
+        }
+
+        Matcher matcher = VERSION_PATTERN.matcher(version);
+        if (matcher.matches()) {
+            try {
+                int calculatedVersion = MAJOR_VERSION_MULTIPLIER * Integer.valueOf(matcher.group(1))
+                        + MINOR_VERSION_MULTIPLIER * Integer.valueOf(matcher.group(2));
+
+                int groupCount = matcher.groupCount();
+                if (groupCount >= PATCH_GROUP_COUNT) {
+                    String patchVersionString = matcher.group(PATCH_GROUP_COUNT);
+                    if (null != patchVersionString && !patchVersionString.startsWith("-")) {
+                        calculatedVersion += Integer.valueOf(patchVersionString);
+                    }
+                }
+                return calculatedVersion;
+            } catch (Exception e) {
+                Logger.getLogger(BuildInfo.class).warning("Failed to calculate version using version string " + version, e);
+            }
+        }
+
+        return UNKNOWN_HAZELCAST_VERSION;
+    }
+
+    public static int getUnknownHazelcastVersion() {
```

Comment:
Ah, that may be one of the visibility checks, okay. I can have a look at this once the PR is merged. This is surely no blocker :)